### ISSUE TITLE
Row Constraints

### ISSUE THREAD

**Issue by paf31** [purescript/purescript#469]
A new type of constraint which can express row equalities between a sum and another row, a la Ermine:

For example:

```
union :: forall r1 r2 r3. (r1 | r2 ~ r3) => { | r1 } -> { | r2 } -> { | r3 }
```

I would be interested to see to what extent something like this could help with expressing interesting constraints on effect sets in `Eff`.

**Comment by LiamGoodacre** [comment]
What (other) operations would we want?
* RowUnion - disjoint union: labels must be distinct across rows
* RowUnionL - non-disjoint union, left biased on overlapping labels
* RowUnionR - non-disjoint union, right biased on overlapping labels
* RowCons - add a field to a row that lacks it
* RowLacks - compute a row without a label
* RowDiff - compute a row without all the labels of another

We'd have to get kind variables in before this.  If we then implement the classes in `Prim` as, for example:
```purescript
class RowUnion (l :: # k) (r :: # k) (o :: # k) | l r -> o
class RowUnionL (l :: # k) (r :: # k) (o :: # k) | l r -> o
class RowUnionR (l :: # k) (r :: # k) (o :: # k) | l r -> o
class RowCons (s :: Symbol) (t :: k) (r :: # k) (o :: # k) | s r -> t o
class RowLacks (s :: Symbol) (r :: # k) (o :: # k) | s r -> o
class RowDiff (l :: # k) (r :: # k) (o :: # k) | l r -> o
```
We wouldn't need to depend on type families/named fundeps, and the type equality constraint.

**Comment by paf31** [comment]
In Ermine, there is just the one constraint, for disjoint partitions, as I understand it. I think you can encode many of these in terms of that one.

**Comment by doolse** [comment]
As an introduction for contributing to the compiler, I implemented a `*` fixed `RowUnionR`:

```purescript 
class TypeRowUnionR (l :: # *) (r :: # *) (o :: # *) | l r -> o
```
which allows for a library (lets say purescript-records) to implement `merge` (or union?):
```purescript
-- | Merge two records together with the second record overriding fields from the first
foreign import merge :: forall a b c. TypeRowUnionR a b c => {|a} -> {|b} -> {|c}
```

I can see a benefit in releasing this before the full row constraints implementation is available (unless it's right around the corner!) as it'll allow for easily creating type safe purescript wrappers around existing Javascript libraries which are based on large amounts of optional properties (React components / React Native etc..) 

Want me to polish up and make a pull request?

**Comment by paf31** [comment]
@doolse It's great that you've got this working, but I think I would like to iron out a few design details before we add row constraints. Specifically, what to do about duplicate labels and kind polymorphism.

Do you have a branch we can see?

**Comment by doolse** [comment]
Here is my branch (rebased on top of the duplicate labels commit):
https://github.com/doolse/purescript/tree/recordunion

There is a passing test for the use case i'm really interested in, (mandatory + optional fields) and a failing test for preventing you from using fields that aren't part of the union.

**Comment by paf31** [comment]
@doolse This looks great! I wonder if we could make the inference work so that any two arguments imply the third. I feel like this should be simpler now with the new duplicate labels rule.

Could you please make a PR so we can go through it there?

Do you have a `purescript-records` repo to go with this?

**Comment by pkamenarsky** [comment]
Would this make it possible to express "extends" constraints on rows?

I.e. in pseudosyntax (this needs record instances, maybe there is another way?):

    class TypeRowExtends super sub

    instance extends :: TypeRowExtends {a|b} {a|c}

or something like that?

I think this would make it possible to make some progress on #1216, if at all possible.

**Comment by paf31** [comment]
Yes, because `r` extends `s` if there exists some other row which unions with `s` to give `r`.

**Comment by chrisdone** [comment]
Just to share a modest approach, for a client DSL I've implemented simply typed lambda calculus + row types, but needed a way to express "JOIN" (as in SQL), which necessitated a type of union for rows. Given the restricted nature of the DSL, I built union types into the type system.

I ended up with row variables being a set of variables instead of one per row, with this type of syntax. So for example,

```haskell
foo row1 row2 = row1 |> JOIN row2 ["key"] ["key"]
```
with `foo` now having the type:

```haskell
foo : { key : t | r1 } -> { key : t | r2 } -> { key : t | r1 + r2 }
```

I didn't have to change the inferrer much in the end.
